EnC: Handle anonymous types and delegates hoisted to closure or state machine

tmat opened the pull request:
Fixes 1170899: fatal error during EnC.
Fixes #3233, #3192.

**Scenarios**
When editing a method multiple times that contains a variable of an anonymous type (or delegate in VB) that is hoisted to a closure or a state machine the debugger reports a fatal error and terminates debugging session. 

Examples:
```VB
    Sub Main()
        Dim x = Function(n As Integer) n + 1
        Dim y = Sub(n As Integer) Console.WriteLine(x(n))
        y(100)
    End Sub
```

```C# 
        static void Main(string[] args)
        {
            var result =
                from a in args
                let x = a.Reverse()
                let y = x.Reverse()
                where x.SequenceEqual(y)
                select new { Value = a, Length = a.Length };            
        }
```

**Fix**
We didn't deeply translate anonymous types/delegates before comparing them in symbol matcher.
The proposed fix for RTM isn't ideal but addresses the issues and avoids risky changes. A proper fix requires extensive changes in symbol translations and should be implemented post RTM.

**Risk**
Although not ideal I'm confident that that fix is addressing the root cause of the issue. The fix introduces new code paths which are all covered by newly added unit tests. These code paths are only exercised during EnC specific code generation and emit and are not taken during regular compilation, which limits the impact only to EnC scenarios.

**Testing**
Added unit tests and manually verified that affected scenarios work for both VB and C#. Buddy tested by @wschae and @amcasey. 

@ManishJayaswal

tmat commented:
@cston @AlekseyTs @wschae @amcasey

cston commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -1316,7 +1316,8 @@ private BoundStatement ChainImplicitStructConstructor(MethodSymbol methodSymbol,
                 {
                     Debug.Assert(method.IsAsync || method.IsIterator);
-                    GetStateMachineSlotDebugInfo(moduleBuilder.GetSynthesizedFields(stateMachineTypeOpt), out stateMachineHoistedLocalSlots, out stateMachineAwaiterSlots);
+                    GetStateMachineSlotDebugInfo(moduleBuilder, moduleBuilder.GetSynthesizedFields(stateMachineTypeOpt), diagnostics, out stateMachineHoistedLocalSlots, out stateMachineAwaiterSlots);
```
Should this be 'diagnosticsForThisMethod'? Or can we remove 'diagnosticsForThisMethod' and use 'diagnostics' everywhere?

tmat replied:
Yes.

cston commented on src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs:
```diff
@@ -752,11 +751,31 @@ public SymbolComparer(MatchSymbols matcher)
+                        right = AnonymousTypeManager.TranslateAnonymousTypeSymbol(anonymousRight);
+                    }
+
+                    if ((object)anonymousLeft != null || (object)anonymousRight != null)
```
Would this get any simpler extracting a helper method?
if (TranslateIfAnonymousTypePublicSymbol(ref left) || TranslateIfAnonymousTypePublicSymbol(ref right)) { ... }

AlekseyTs replied:
Should we be using ```&&``` instead?

tmat replied:
No. One of the symbols may be public symbol the other may be PE symbol of the implementation.

cston commented on src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs:
```diff
@@ -96,7 +96,7 @@ private FieldSymbol GetAwaiterField(TypeSymbol awaiterType)
             {
                 int slotIndex;
-                if (slotAllocatorOpt == null || !slotAllocatorOpt.TryGetPreviousAwaiterSlotIndex((Cci.ITypeReference)awaiterType, out slotIndex))
+                if (slotAllocatorOpt == null || !slotAllocatorOpt.TryGetPreviousAwaiterSlotIndex(F.ModuleBuilderOpt.Translate(awaiterType, null, F.Diagnostics), out slotIndex))
```
Is there an assert that could be added to TryGetPreviousAwaiterSlotIndex and TryGetPreviousHoistedLocalSlotIndex to catch untranslated types?

cston replied:
Pass F.Syntax for syntax argument.

tmat replied:
Re assert: for most types translated types are the same instance. We would need to check for specific translated types (like NoPIA, anonymous, etc.) But we can't currently do that easily in language agnostic way :(

tmat replied:
F.Syntax ... Done.

tmat replied:
Post RTM we could probably introduce struct TranslatedSymbol<T> and have the type system to check it for us.

cston commented on src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs:
```diff
@@ -752,11 +751,31 @@ public SymbolComparer(MatchSymbols matcher)
+                    }
+
+                    var anonymousLeft = left as AnonymousTypeManager.AnonymousTypePublicSymbol;
+                    if ((object)anonymousLeft != null)
```
if (left.IsAnonymousType)

tmat replied:
That would require left to be TypeSymbol, which it is not.

cston commented on src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousType.TypePublicSymbol.cs:
```diff
@@ -22,7 +22,7 @@ internal sealed partial class AnonymousTypeManager
         /// In emit phase it is being substituted with implementation symbol.
         /// </summary>
-        private sealed class AnonymousTypePublicSymbol : NamedTypeSymbol
+        internal sealed class AnonymousTypePublicSymbol : NamedTypeSymbol
```
Does this need to be internal if SymbolMatcher uses .IsAnonymousType?

tmat replied:
Perhaps not, but the code would be more complex. Also the corresponding type is already internal in VB.

cston commented on src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType.PropertySymbol.cs:
```diff
@@ -19,7 +19,7 @@ internal sealed partial class AnonymousTypeManager
         /// Represents an anonymous type template's property symbol.
         /// </summary>
-        private sealed class AnonymousTypePropertySymbol : PropertySymbol
+        internal sealed class AnonymousTypePropertySymbol : PropertySymbol
```
Is it necessary to make AnonymousTypePropertySymbol, AnonymousTypeTemplateSymbol, and NameAndIndex internal?

tmat replied:
Once AnonymousTypePublicSymbol is the other classes need to as well, since they are used by non-private members of AnonymousTypePublicSymbol.

cston commented on src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicDefinitionMap.vb:
```diff
@@ -121,10 +123,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
 
                             If GeneratedNames.TryParseSlotIndex(StringConstants.StateMachineAwaiterFieldPrefix, name, slotIndex) Then
-                                Dim field = TryCast(member, IFieldSymbol)
+                                Dim field = DirectCast(member, IFieldSymbol)
```
Can we use FieldSymbol rather than IFieldSymbol?

cston commented:
LGTM

AlekseyTs commented on src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs:
```diff
@@ -653,7 +653,6 @@ private static MethodSymbol SubstituteTypeParameters(MethodSymbol method)
 
             private bool AreNamedTypesEqual(NamedTypeSymbol type, NamedTypeSymbol other)
```
I find the name of this method very confusing. It looks like its logic is specific to Anonymous Types, perhaps the name should reflect this and asserting on the input night be appropriate.

AlekseyTs replied:
Names of the members are significant for Anonymous Types. Should we compare them?

tmat replied:
The method is not specific to anon types.

Yes, good point.

wschae commented on src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueClosureTests.cs:
```diff
@@ -2625,5 +2625,148 @@ .maxstack  3
+                ImmutableArray.Create(
+                    new SemanticEdit(SemanticEditKind.Update, f1, f2, GetSyntaxMapFromMarkers(source1, source2), preserveLocalVariables: true)));
+
+            diff1.VerifySynthesizedMembers(
```
diff2?

tmat replied:
Yes. Thanks.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb:
```diff
@@ -510,7 +510,6 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
             End Function
 
             Private Function AreNamedTypesEqual(type As NamedTypeSymbol, other As NamedTypeSymbol) As Boolean
```
Comments for C# version of this function is applicable here as well. In addition, it doesn't look like this function makes any distinction between Anonymous Types and Anonymous Delegates. It feel like the distinction should be made.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Emit/EditAndContinue/VisualBasicSymbolMatcher.vb:
```diff
@@ -587,8 +586,26 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Emit
+                        right = anonymousRight.MapToImplementationSymbol()
+                    End If
+
+                    If anonymousLeft IsNot Nothing OrElse anonymousRight IsNot Nothing Then
```
AndAlso?

wschae commented on src/Compilers/VisualBasic/Test/Emit/Emit/EditAndContinue/EditAndContinueTestBase.vb:
```diff
@@ -21,6 +22,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests
 
         Friend Shared ReadOnly EmptyLocalsProvider As Func(Of MethodDefinitionHandle, EditAndContinueMethodDebugInformation) = Function(token) Nothing
 
+        Friend Shared Function Visualize(baseline As ModuleMetadata, ParamArray deltas As PinnedMetadata()) As String
```
Is it for future reference? I don't see anyone calling it.

tmat replied:
Yes, it's a helper one can use to diagnose issues.

tmat commented:
@cston @AlekseyTs I have revisited the implementation of comparer in SymbolMatcher, hydration of PE symbols from initial metadata in PEDeltaAssemblyBuilder and added more tests. Please, review the updated change.

cston commented on src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs:
```diff
@@ -752,11 +759,41 @@ public SymbolComparer(MatchSymbols matcher)
+
+                        // Map the template - this involves comparing the structure of the anonymous type,
+                        // so we don't compare equal types with different field names.
+                        if (_matcher.Visit(namedLeft.ConstructedFrom)?.Equals(namedRight.ConstructedFrom) != true)
```
Would == be sufficient? if (_matcher.Visit(namedLeft.ConstructedFrom) == namedRight.ConstructedFrom) { ... }

tmat replied:
Yes.

AlekseyTs replied:
Is it possible to get into a cycle here (ConstructedFrom can return the same symbol back)?

cston commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -1316,7 +1316,8 @@ private BoundStatement ChainImplicitStructConstructor(MethodSymbol methodSymbol,
                     Debug.Assert(method.IsAsync || method.IsIterator);
-                    GetStateMachineSlotDebugInfo(moduleBuilder.GetSynthesizedFields(stateMachineTypeOpt), out stateMachineHoistedLocalSlots, out stateMachineAwaiterSlots);
+                    GetStateMachineSlotDebugInfo(moduleBuilder, moduleBuilder.GetSynthesizedFields(stateMachineTypeOpt), diagnosticsForThisMethod, out stateMachineHoistedLocalSlots, out stateMachineAwaiterSlots);
+                    Debug.Assert(!diagnostics.HasAnyErrors());
```
diagnosticsForThisMethod

cston commented:
LGTM

wschae commented on src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs:
```diff
@@ -343,6 +349,7 @@ public override Symbol Visit(Symbol symbol)
 
                 // Add an entry for the match, even if there is no match, to avoid
                 // matching the same symbol unsuccessfully multiple times.
+                var x = base.Visit(symbol);
```
Is it for debugging purpose?

tmat replied:
Yes, forgot to update the PR. I have removed it.

MattGertz commented:
MoreInfo 6/1 by ML Shiproom.  Please get buddy-testing on this.  It clearly meets the bar, but we just want to make sure that we have no regressions because this is a big change touching a lot of files.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType.TemplateSymbol.cs:
```diff
@@ -133,7 +133,7 @@ internal AnonymousTypeTemplateSymbol(AnonymousTypeManager manager, AnonymousType
             internal AnonymousTypeKey GetAnonymousTypeKey()
             {
-                var properties = this.Properties.SelectAsArray(p => AnonymousTypeKeyField.CreateField(p.Name));
+                var properties = this.Properties.SelectAsArray(p => new AnonymousTypeKeyField(p.Name, isKey: false, ignoreCase: false));
```
What is the reason for this change? Please revert unless behavior has changed.

tmat replied:
AnonymousTypeKeyField did convert the name to lower-case (when it was created from VB) instead of comparing case-insensitively. Lowering case and then comparing ordinal is incorrect.

AlekseyTs commented:
LGTM

dpoeschl commented:
retest this please

tmat commented:
@AlekseyTs @cston @amcasey @wschae @ManishJayaswal Figured out a fairly low risk workaround for the anonymous type mapping issues. Implemented as a separate commit on top of the previous PR. All issues that I know of are now addressed. Please, review as soon as tests pass.

AlekseyTs commented:
All this magic around deep translation can benefit from detailed comments about why and what are we doing.

tmat commented:
@AlekseyTs I'll add a comment to make it more clear.

cston commented:
LGTM

tmat commented:
@AlekseyTs I have removed the direct cats to ITypeReference and added the comment.

AlekseyTs commented:
LGTM

wschae commented:
Buddy Tested: 

- verified EnC not working w/o PR
- verified EnC working with PR
- intensively covered a combination usage of anonymous types, lambdas, async and iterators
- checked both C# and VB

LGTM

amcasey commented:
I'm out of buddy testing ideas and it's been more than an hour since I broke anything EnC related. :)

:+1:

pharring commented:
Approved 6/5 ML Shiproom.
@tmat Please add an issue to do the "right fix" to the compiler in Milestone 1.1

tmat commented:
Filed workitem #3341
